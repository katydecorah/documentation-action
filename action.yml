name: GitHub Action Documentation
author: "@katydecorah"
description: "A GitHub action to help document your GitHub action."
runs:
  using: "node16"
  main: dist/index.js

inputs:
  exampleWorkflowFile:
    description: The example workflow file in `.github/workflows/`
    default: example.yml
  additionalWorkflowFilePrefix:
    description: To include additional workflow files, save them with this prefix in `.github/workflow/`.
    default: example
  documentationFile:
    description: The file where the action will write and update documentation for the action.
    default: README.md
