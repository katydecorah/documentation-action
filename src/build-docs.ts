import { formatInputs } from "./format-inputs";

export function buildDocs({ exampleWorkflowYaml, action, release }) {
  let docs = `
## Set up the workflow

\`\`\`yml
${trimExampleWorkflow({ exampleWorkflowYaml, release })}
\`\`\`
`;
  // Document inputs, if they exist
  if ("inputs" in action) {
    docs += formatInputs(action.inputs);
  }
  return docs;
}

export function trimExampleWorkflow({ exampleWorkflowYaml, release }) {
  return exampleWorkflowYaml.replace("uses: ./", `uses: ${release}`);
}
