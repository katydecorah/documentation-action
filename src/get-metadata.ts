import { readFile, readdir } from "fs/promises";
import { load } from "js-yaml";
import { getInput } from "@actions/core";
import { ActionConfig } from "./action";

export async function getWorkflow(): Promise<WorkflowConfig | undefined> {
  const exampleWorkflowFile: string = getInput("example-workflow-file");
  try {
    const yaml = await readFile(
      `./.github/workflows/${exampleWorkflowFile}`,
      "utf-8"
    );
    return {
      yaml,
      json: load(yaml) as WorkflowJson,
    };
  } catch (error) {
    throw new Error(error);
  }
}

export async function getWorkflows(): Promise<WorkflowConfig[] | undefined> {
  const exampleWorkflowFile: string = getInput("example-workflow-file");
  const additionalWorkflowFilePrefix: string = getInput(
    "additional-workflow-file-prefix"
  );
  try {
    const workflows = await readdir("./.github/workflows/");
    const additionalWorkflows = workflows.filter(
      (f) =>
        f.startsWith(additionalWorkflowFilePrefix) && f !== exampleWorkflowFile
    );

    if (additionalWorkflows.length === 0) return undefined;

    const workflowArray: WorkflowConfig[] = [];

    for (const workflow of additionalWorkflows) {
      const yaml = await readFile(`./.github/workflows/${workflow}`, "utf-8");
      workflowArray.push({
        yaml,
        json: load(yaml) as WorkflowJson,
      });
    }
    return workflowArray;
  } catch (error) {
    throw new Error(error);
  }
}

export async function getActionConfig(): Promise<ActionConfig> {
  try {
    const actionConfig = await readFile("./action.yml", "utf-8");
    return load(actionConfig) as ActionConfig;
  } catch (error) {
    throw new Error(error);
  }
}

export async function getRelease(): Promise<string> {
  try {
    const { version } = JSON.parse(await readFile("./package.json", "utf-8"));
    return `${process.env.GITHUB_REPOSITORY}@v${version}`;
  } catch (error) {
    throw new Error(error);
  }
}

export type WorkflowConfig = {
  yaml: string;
  json: WorkflowJson;
};

export type Input = {
  description: string;
  required?: boolean;
  type?: string;
  default?: string;
  deprecationMessage?: string;
  options?: string[];
};

export type Inputs = {
  [key: string]: Input;
};

export type WorkflowJson = {
  name: string;
  on: {
    push: {
      paths: string[];
    };
    workflow_dispatch?: {
      inputs?: Inputs;
    };
  };
  jobs: Record<string, unknown>;
};
