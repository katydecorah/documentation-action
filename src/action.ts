import { setFailed } from "@actions/core";
import { writeDocs } from "./write-docs";
import { buildDocs } from "./build-docs";
import { getWorkflow, getActionConfig, getRelease } from "./get-metadata";

export async function docs() {
  try {
    // Get workflow metadata
    const [exampleWorkflowYaml, action, release] = await Promise.all([
      getWorkflow(),
      getActionConfig(),
      getRelease(),
    ]);

    if (!exampleWorkflowYaml || !action || !release) {
      setFailed("Unable to get action metadata");
      return;
    }
    // Build docs
    const docs = buildDocs({ exampleWorkflowYaml, action, release });
    // Write docs
    await writeDocs(docs);
  } catch (error) {
    setFailed(error.message);
  }
}

/**
 * As defined by:
 * https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions
 */
export type ActionConfig = {
  name: string;
  author: string;
  description: string;
  runs: {
    using: string;
    main: string;
  };
  inputs: {
    /** A string identifier to associate with the input. */
    [input: string]: {
      /** A string description of the input parameter. */
      description: string;
      /** A boolean to indicate whether the action requires the input parameter. */
      required: boolean;
      /** A string representing the default value. */
      default?: string;
      /** If the input parameter is used, this string is logged as a warning message. */
      deprecationMessage?: string;
    };
  };
};
