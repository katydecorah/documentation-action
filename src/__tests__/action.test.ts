import { ActionConfig, docs } from "../action";
import { readFileSync, promises } from "fs";
import * as GetMetadata from "../get-metadata";
import * as BuildDocs from "../build-docs";
import { load } from "js-yaml";
import { setFailed } from "@actions/core";
import * as core from "@actions/core";

jest.mock("@actions/core");

const action = load(readFileSync("./action.yml", "utf-8")) as ActionConfig;
const workflowYaml = readFileSync("./.github/workflows/example.yml", "utf-8");

const workflow = {
  yaml: workflowYaml,
  json: load(workflowYaml),
};

beforeEach(() => {
  jest.spyOn(core, "getInput").mockImplementation(() => "README.md");
});

describe("action", () => {
  test("works", async () => {
    const writeFileSpy = jest.spyOn(promises, "writeFile").mockImplementation();
    const getWorkflowSpy = jest
      .spyOn(GetMetadata, "getWorkflow")
      .mockReturnValueOnce(Promise.resolve(workflow));
    const getActionSpy = jest
      .spyOn(GetMetadata, "getActionConfig")
      .mockReturnValueOnce(Promise.resolve(action));
    const getRelease = jest
      .spyOn(GetMetadata, "getRelease")
      .mockReturnValueOnce(
        Promise.resolve(`katydecorah/documentation-action@v1.0.0`)
      );
    const buildDocsSpy = jest.spyOn(BuildDocs, "buildDocs");

    await docs();
    expect(getWorkflowSpy).toHaveBeenCalled();
    expect(getActionSpy).toHaveBeenCalled();
    expect(getRelease).toHaveBeenCalled();
    expect(buildDocsSpy).toHaveBeenCalled();
    expect(buildDocsSpy.mock.results[0].value).toMatchSnapshot();
    expect(writeFileSpy.mock.calls[0]).toMatchSnapshot();
  });

  test("writeFile fails", async () => {
    jest
      .spyOn(GetMetadata, "getWorkflow")
      .mockReturnValueOnce(Promise.resolve(workflow));
    jest
      .spyOn(GetMetadata, "getActionConfig")
      .mockReturnValueOnce(Promise.resolve(action));
    jest
      .spyOn(GetMetadata, "getRelease")
      .mockReturnValueOnce(
        Promise.resolve(`katydecorah/documentation-action@v1.0.0`)
      );
    jest.spyOn(BuildDocs, "buildDocs");

    jest.spyOn(promises, "writeFile").mockRejectedValue({ message: "Error" });

    await docs();
    expect(setFailed).toHaveBeenCalledWith({ message: "Error" });
  });

  test("get metadata fails", async () => {
    jest
      .spyOn(GetMetadata, "getWorkflow")
      .mockReturnValueOnce(Promise.resolve(undefined));
    jest
      .spyOn(GetMetadata, "getActionConfig")
      .mockReturnValueOnce(Promise.resolve(undefined));
    jest
      .spyOn(GetMetadata, "getRelease")
      .mockReturnValueOnce(Promise.resolve(undefined));

    await docs();
    expect(setFailed).toHaveBeenCalledWith("Unable to get action metadata");
  });
});
