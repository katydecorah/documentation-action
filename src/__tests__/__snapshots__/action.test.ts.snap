// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action works 1`] = `
"
## Set up the workflow

To use this action, create a new workflow in \`.github/workflows\` and modify it as needed:

\`\`\`yml
name: Document GitHub action

on:
  push:
    paths:
      - ".github/workflows/example.yml"
      - "action.yml"
      - "package.json"
      - "README.md"

jobs:
  document_action:
    runs-on: macOS-latest
    name: Write documentation
    steps:
      - name: Checkout
        uses: actions/checkout@v3
      - name: Documentation action
        uses: katydecorah/documentation-action@v1.0.0
      - name: Commit files
        if: env.UpdateDocumentation == 'true'
        run: |
          git pull
          git config --local user.email "action@github.com"
          git config --local user.name "GitHub Action"
          git commit -am "Update documentation"
          git push
\`\`\`

## Action options

- \`exampleWorkflowFile\`: The example workflow file in \`.github/workflows/\` Default: \`example.yml\`.
- \`additionalWorkflowFilePrefix\`: To include additional workflow files, save them with this prefix in \`.github/workflow/\`. Default: \`example\`.
- \`documentationFile\`: The file where the action will write and update documentation for the action. Default: \`README.md\`.
"
`;

exports[`action works 2`] = `
[
  "./README.md",
  "# documentation-action

A GitHub action to help document your GitHub actions.

To use this action, create two workflow files:

1. \`.github/workflows/example.yml\` - in this file, create an example action to show developers how to use it.
2. \`.github/workflows/documentation.yml\` - this will hold handle \`documentation-action\` which is documented below. Follow the steps below to set up the documetnation action:

<!-- START GENERATED DOCUMENTATION -->

## Set up the workflow

To use this action, create a new workflow in \`.github/workflows\` and modify it as needed:

\`\`\`yml
name: Document GitHub action

on:
  push:
    paths:
      - ".github/workflows/example.yml"
      - "action.yml"
      - "package.json"
      - "README.md"

jobs:
  document_action:
    runs-on: macOS-latest
    name: Write documentation
    steps:
      - name: Checkout
        uses: actions/checkout@v3
      - name: Documentation action
        uses: katydecorah/documentation-action@v1.0.0
      - name: Commit files
        if: env.UpdateDocumentation == 'true'
        run: |
          git pull
          git config --local user.email "action@github.com"
          git config --local user.name "GitHub Action"
          git commit -am "Update documentation"
          git push
\`\`\`

## Action options

- \`exampleWorkflowFile\`: The example workflow file in \`.github/workflows/\` Default: \`example.yml\`.
- \`additionalWorkflowFilePrefix\`: To include additional workflow files, save them with this prefix in \`.github/workflow/\`. Default: \`example\`.
- \`documentationFile\`: The file where the action will write and update documentation for the action. Default: \`README.md\`.
<!-- END GENERATED DOCUMENTATION -->
",
]
`;

exports[`action works, workflows 1`] = `
"
## Set up the workflow

To use this action, create a new workflow in \`.github/workflows\` and modify it as needed:

\`\`\`yml
name: Document GitHub action

on:
  push:
    paths:
      - ".github/workflows/example.yml"
      - "action.yml"
      - "package.json"
      - "README.md"

jobs:
  document_action:
    runs-on: macOS-latest
    name: Write documentation
    steps:
      - name: Checkout
        uses: actions/checkout@v3
      - name: Documentation action
        uses: katydecorah/documentation-action@v1.0.0
      - name: Commit files
        if: env.UpdateDocumentation == 'true'
        run: |
          git pull
          git config --local user.email "action@github.com"
          git config --local user.name "GitHub Action"
          git commit -am "Update documentation"
          git push
\`\`\`

 ### Additional example workflows

<details>
<summary>Document GitHub action (advanced)</summary>

\`\`\`yml
name: Document GitHub action (advanced)
# This workflow file is the same as example.yml
# It's a proof of concept that you provide additional workflow files and they will appear in the README.

jobs:
  document_action:
    runs-on: macOS-latest
    name: Write documentation
    steps:
      - name: Checkout
        uses: actions/checkout@v3
      - name: Documentation action
        uses: katydecorah/documentation-action@v1.0.0
      - name: Commit files
        if: env.UpdateDocumentation == 'true'
        run: |
          git pull
          git config --local user.email "action@github.com"
          git config --local user.name "GitHub Action"
          git commit -am "Update documentation"
          git push
\`\`\`

</details>

<details>
<summary>Document GitHub action (new feature)</summary>

\`\`\`yml
name: Document GitHub action (new feature)
# This workflow file is the same as example.yml
# It's a proof of concept that you provide additional workflow files and they will appear in the README.

jobs:
  document_action:
    runs-on: macOS-latest
    name: Write documentation
    steps:
      - name: Checkout
        uses: actions/checkout@v3
      - name: Documentation action
        uses: katydecorah/documentation-action@v1.0.0
      - name: Commit files
        if: env.UpdateDocumentation == 'true'
        run: |
          git pull
          git config --local user.email "action@github.com"
          git config --local user.name "GitHub Action"
          git commit -am "Update documentation"
          git push
\`\`\`

</details>


## Action options

- \`exampleWorkflowFile\`: The example workflow file in \`.github/workflows/\` Default: \`example.yml\`.
- \`additionalWorkflowFilePrefix\`: To include additional workflow files, save them with this prefix in \`.github/workflow/\`. Default: \`example\`.
- \`documentationFile\`: The file where the action will write and update documentation for the action. Default: \`README.md\`.
"
`;

exports[`action works, workflows 2`] = `
[
  "./README.md",
  "# documentation-action

A GitHub action to help document your GitHub actions.

To use this action, create two workflow files:

1. \`.github/workflows/example.yml\` - in this file, create an example action to show developers how to use it.
2. \`.github/workflows/documentation.yml\` - this will hold handle \`documentation-action\` which is documented below. Follow the steps below to set up the documetnation action:

<!-- START GENERATED DOCUMENTATION -->

## Set up the workflow

To use this action, create a new workflow in \`.github/workflows\` and modify it as needed:

\`\`\`yml
name: Document GitHub action

on:
  push:
    paths:
      - ".github/workflows/example.yml"
      - "action.yml"
      - "package.json"
      - "README.md"

jobs:
  document_action:
    runs-on: macOS-latest
    name: Write documentation
    steps:
      - name: Checkout
        uses: actions/checkout@v3
      - name: Documentation action
        uses: katydecorah/documentation-action@v1.0.0
      - name: Commit files
        if: env.UpdateDocumentation == 'true'
        run: |
          git pull
          git config --local user.email "action@github.com"
          git config --local user.name "GitHub Action"
          git commit -am "Update documentation"
          git push
\`\`\`

 ### Additional example workflows

<details>
<summary>Document GitHub action (advanced)</summary>

\`\`\`yml
name: Document GitHub action (advanced)
# This workflow file is the same as example.yml
# It's a proof of concept that you provide additional workflow files and they will appear in the README.

jobs:
  document_action:
    runs-on: macOS-latest
    name: Write documentation
    steps:
      - name: Checkout
        uses: actions/checkout@v3
      - name: Documentation action
        uses: katydecorah/documentation-action@v1.0.0
      - name: Commit files
        if: env.UpdateDocumentation == 'true'
        run: |
          git pull
          git config --local user.email "action@github.com"
          git config --local user.name "GitHub Action"
          git commit -am "Update documentation"
          git push
\`\`\`

</details>

<details>
<summary>Document GitHub action (new feature)</summary>

\`\`\`yml
name: Document GitHub action (new feature)
# This workflow file is the same as example.yml
# It's a proof of concept that you provide additional workflow files and they will appear in the README.

jobs:
  document_action:
    runs-on: macOS-latest
    name: Write documentation
    steps:
      - name: Checkout
        uses: actions/checkout@v3
      - name: Documentation action
        uses: katydecorah/documentation-action@v1.0.0
      - name: Commit files
        if: env.UpdateDocumentation == 'true'
        run: |
          git pull
          git config --local user.email "action@github.com"
          git config --local user.name "GitHub Action"
          git commit -am "Update documentation"
          git push
\`\`\`

</details>


## Action options

- \`exampleWorkflowFile\`: The example workflow file in \`.github/workflows/\` Default: \`example.yml\`.
- \`additionalWorkflowFilePrefix\`: To include additional workflow files, save them with this prefix in \`.github/workflow/\`. Default: \`example\`.
- \`documentationFile\`: The file where the action will write and update documentation for the action. Default: \`README.md\`.
<!-- END GENERATED DOCUMENTATION -->
",
]
`;
